src/coreclr/gc/gc.cpp

```diff
@@ -17923,7 +17927,7 @@ BOOL gc_heap::allocate_more_space(alloc_context* acontext, size_t size,
             if (heap_hard_limit && (status == a_state_retry_allocate))
             {
                 alloc_heap = balance_heaps_uoh_hard_limit_retry (acontext, size, alloc_generation_number);
-                if (alloc_heap == nullptr)
+                if (alloc_heap == nullptr || retry_count++ == ALLOCATION_RETRY_MAX_COUNT)
```

Comment:
nit - in GC code we put parentheses around each sub expression so I would write this line as 
`if ((alloc_heap == nullptr) || (retry_count++ == ALLOCATION_RETRY_MAX_COUNT))`

also it's probably better to name this const UOH_ALLOC_RETRY_MAX_COUNT since this only applies to UOH